### PR TITLE
[DLinkSmartHome] add DSP-W215 Smart Plug

### PR DESCRIPTION
**Pull request by pasbi**
Note: the original PR (2.5.x) is  #8361.

## DESCRIPTION

- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
Add support for a thing.
- Did you describe the PRs motivation and goal?
Read temperature, state and total/current power consumption of the thing, set the state of the thing.
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
No.
- Did you describe new features for the end user?
Yes, in the README.md
- Did you describe any noteworthy changes in usage for the end user?
Not necessary.
- Was the documentation updated accordingly, e.g. the add-on README?
Yes.
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
It passed all the checks, and I'm not aware of any violation. Yet, I'm not sure. BTW, the link is broken.
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
Yes.
- Did you sign-off your work:
Yes.

## TESTING

### CODE REVIEW

File: bundles/org.openhab.binding.dlinksmarthome/pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
-    <version>3.0.0-SNAPSHOT</version>
+    <version>2.5.9-SNAPSHOT</version>
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/pom.xml:10]: Is this change intentional?

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+/**
+ * The {@link DLinkHNAP} class encapsulates the HNAP implementation for DLink thing handlers.
+ *
+ * @author Pascal Bies - Initial contribution
+ */
+public abstract class DLinkHNAP {
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:66]: Please add the NonNullByDefault annotation.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        }
+
+        public String privateKey;
+        public String cookie;
+    }
+
+    public void start(ScheduledExecutorService scheduler) {
+        try {
+            pollFuture = scheduler.scheduleWithFixedDelay(this::poller, 0, DETECT_POLL_S, TimeUnit.SECONDS);
+        } catch (Exception e) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:118]: What type of exception do you expect here? Can you specify the concrete type?

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR,
+                    "Unexpected internal error.");
+            logger.error("Failed to start http client or scheduler.");
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:121]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be warn.

Additionally, you could remove the logging statement at all, as `updateStatus()` already does the logging.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        try {
+            final Document document = post(action, newMessage(action, moduleParameters(module)), authentication);
+            return getElementByTagName(document, name);
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:168]: You could include the exceptions message as third parameter to `updateStatus()` to display the error in the UI.
You could remove the logging statement, as `updateStatus()` already does the logging.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    }
+
+    public State queryOnOffState(final String action, final String name, final String module) {
+        final String state = queryStringState(action, name, module);
+        if ("true".contentEquals(state)) {
+            return OnOffType.ON;
+        } else if ("false".equals(state)) {
+            return OnOffType.OFF;
+        } else {
+            logger.error("Unexpected state: {}", state);
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:179]: See above. Logging to error. Please take care of all occurences.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                    soapFactory.createElement("Description").addTextNode(description),
+                    soapFactory.createElement("OPStatus").addTextNode(state));
+            final Document document = post(action, message, authentication);
+            final String response = getElementByTagName(document, name);
+            if (!"ok".equals(response.toLowerCase())) {
+                logger.warn("Failed to set state.");
+            }
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:217]: See above. Set ThingStatusDetail.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+        } catch (Exception e) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:218]: What type of exception do you expect here? Can you specify the concrete type?

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        } catch (CommunicationException e) {
+            authentication = null;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Failed to authenticate.");
+            return false;
+        } catch (SOAPException e) {
+            authentication = null;
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+            return false;
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:299]: See above. Include the error message. Remove logging. Please see the other occurences below.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    }
+
+    private Document post(final String action, final SOAPMessage message) throws CommunicationException {
+        return post(action, message, null);
+    }
+
+    private Document post(final String action, final SOAPMessage message, final Authentication authentication)
+            throws CommunicationException {
+        Request request = httpClient.newRequest(url);
+        request.timeout(5, TimeUnit.SECONDS);
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:309]: Is 5 sec. sufficient under all circumstances? Most bindings use a timeout of 10 sec.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+
+    private static String toHexString(byte[] bytes) {
+        Formatter formatter = new Formatter();
+        for (byte b : bytes) {
+            formatter.format("%02X", b);
+        }
+        final String hmac = formatter.toString();
+        formatter.close();
+        return hmac;
+    }
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java:376]: You could use the built-in converter: https://www.openhab.org/javadoc/v2.5/org/eclipse/smarthome/core/util/hexutils

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java [hunk 12]
```diff
@@ -34,9 +34,10 @@
  * The {@link DLinkSmartHomeDiscoveryParticipant} is responsible for discovering devices through UPnP.
  *
  * @author Mike Major - Initial contribution
+ * @author Pascal Bies - Add DSP-W215 discovery
  *
  */
-@Component
+@Component(immediate = true)
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java:40]: Is there a reason why you added this?

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java [hunk 13]
```diff
@@ -62,8 +73,17 @@ public DiscoveryResult createResult(final ServiceInfo serviceInfo) {
...
+            final String host = serviceInfo.getHostAddresses()[0];
+            final String mac = serviceInfo.getPropertyString("mac");
+
+            final Map<String, Object> properties = new HashMap<>();
+            properties.put(IP_ADDRESS, host);
+
+            logger.debug("{} found: {}", thingUID.getAsString(), host);
+
+            return DiscoveryResultBuilder.create(thingUID).withThingType(thingTypeUID).withProperties(properties)
+                    .withLabel(thingName(thingTypeUID) + " (" + mac + ")").build();
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java:86]: Can you add the representation property? E.g. the MAC.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java [hunk 14]
```diff
@@ -62,8 +73,17 @@ public DiscoveryResult createResult(final ServiceInfo serviceInfo) {
 
         final ThingTypeUID thingTypeUID = getThingType(serviceInfo);
 
-        if (THING_TYPE_DCHS150.equals(thingTypeUID)) {
-            return createMotionSensor(thingUID, thingTypeUID, serviceInfo);
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java:66]: Can you add a comment to the PR description why you removed this device?

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+            @Override
+            public void handleCommandAuthenticated(ChannelUID channelUID, Command command) {
+                if (command == RefreshType.REFRESH) {
+                    updateState(channelUID, getNewChannelState(channelUID.getId()));
+                } else if (channelUID.equals(getChannelUID(DLinkSmartHomeBindingConstants.STATE))) {
+                    try {
+                        DLinkSmartPlugHandler.this.setState((OnOffType) command);
+                    } catch (ClassCastException e) {
+                        logger.error("Unexpected command type for channel '{}'.", DLinkSmartHomeBindingConstants.STATE);
+                    }
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java:68]: Unchecked exceptions should be cought only if necessary. Please check the type beforehand.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+            protected void updateStatus(ThingStatus status, ThingStatusDetail detail, String message) {
+                DLinkSmartPlugHandler.this.updateStatus(status, detail, message);
+            }
+        };
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        hnap.stop();
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java:82]: The super destructor should be called at the end of the implementing method.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    private void setState(OnOffType state) {
+        hnap.setState("SetSocketSettings", "SetSocketSettingsResult", "1", "Socket 1", "Socket 1", state);
+    }
+
+    private ChannelUID getChannelUID(final String name) {
+        Channel channel = thing.getChannel(name);
+        if (channel == null) {
+            logger.error("Did not find channel '{}'.", name);
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java:127]: See above. this could be warn.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+                    return getCurrentConsumption();
+                case DLinkSmartHomeBindingConstants.TOTAL_CONSUMPTION:
+                    return getTotalConsumption();
+                case DLinkSmartHomeBindingConstants.TEMPERATURE:
+                    return getTemperature();
+                case DLinkSmartHomeBindingConstants.STATE:
+                    return getState();
+                default:
+                    logger.error("Unexpected channel: '{}'.", channelID);
+                    throw new IllegalArgumentException("Unexpected channel.");
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/handler/DLinkSmartPlugHandler.java:147]: Uncought unchecked exceptions are logged by the framework. You could remove the logging message.

File: bundles/org.openhab.binding.dlinksmarthome/src/main/resources/OH-INF/thing/channels.xml [hunk 19]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="dlinksmarthome"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<channel-type id="current_consumption" advanced="true">
```
Comment (fwolter) [inline bundles/org.openhab.binding.dlinksmarthome/src/main/resources/OH-INF/thing/channels.xml:7]: Is there are reason why you made all Channels advanced?

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java (hunk 14 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkSmartHomeDiscoveryParticipant.java:66]: It's not removed. It's just moved to the set part of `SUPPORTED_THING_TYPES_UIDS`

**Comment by Hilbrand** [comment]
Can you also address the comments in #8361 that are shown under `8 hidden conversations`. Also note that in this pr are hidden conversations.

**Comment by fwolter** [comment]
@pasbi What is the status of this PR?

**Comment by pasbi** [comment]
Sorry -- I'm totally confused right now. Please be patient with me.

That confusion comes from two points:
1. The automatic conversion from 2.5.x to 3.x (maybe this is easy, but for me it's intransparent)
2. The fact that the original PR #8361 was kind of rejected (reason: new code does not integrate well into the old one).

So my proposal was to close this PR and start over again with the approach proposed in #8361 :
1. refactor existing code
2. add the new device properly

I'm willing to do that, but I need some advice from the code owner @MikeJMajor to refactor.

**Comment by fwolter** [comment]
I'm not quite sure what you are confused about. You already managed to migrate the code from 2.5 to 3.x. Next step would be to incorporate the review feedback, I gave in this PR.

Maybe you could ask a concrete question, that @MikeJMajor or anybody else can answer?

**Comment by pasbi** [comment]
I'm planning to continue working on this PR, but I'm quite busy atm, unfortunately. If others want to contribute, go ahead :smile:

**Comment by wborn** [comment]
Feel free to reopen this PR when you continue working on it.
